cds_ils/importer/importer.py

```diff
@@ -154,82 +117,118 @@ def index_all_records(self):
+
     def import_record(self):
         """Import record."""
+        document, eitem, series = None, None, None
+        action = None
         self._validate_provider()
 
-        # finds the exact match, update records
-        matched_document = self._match_document()
-
-        if matched_document:
-            self.output_pid = matched_document["pid"]
-            self.action = "update"
-            self.update_records(matched_document)
-            self.index_all_records()
-            return self.import_summary()
-
+        exact_match, partial_matches = self._match_document()
         # finds the multiple matches or fuzzy matches, does not create new doc
         # requires manual intervention, to avoid duplicates
-        if self.ambiguous_matches:
-            return self.import_summary()
-
-        self.document = self.document_importer.create_document()
-        if self.document:
-            self.output_pid = self.document["pid"]
-            self.action = "create"
-            self.eitem_importer.create_eitem(self.document)
-            self.eitem_summary = self.eitem_importer.summary()
-            self.series_list = self.series_importer.import_series(
-                self.document)
-            self.index_all_records()
-        return self.import_summary()
+        partial_matches = self.find_partial_matches(partial_matches)
+
+        # finds the exact match, update records
+        if exact_match:
+            document, eitem, series = self.update_exact_match(exact_match)
+            self.index_records(document, eitem, series)
+            return self.report(document=document,
+                               action="update",
+                               partial_matches=partial_matches,
+                               eitem=eitem, series=series)
+
+        document = self.document_importer.create_document()
+        if document:
+            action = "create"
+            self.eitem_importer.create_eitem(document)
+            eitem = self.eitem_importer.summary()
+            series = self.series_importer.import_series(document)
+            self.index_records(document, eitem, series)
+
+        return self.report(document=document,
+                           action=action,
+                           partial_matches=partial_matches,
+                           eitem=eitem, series=series)
 
     def delete_record(self):
         """Deletes the eitems of the record."""
         document_indexer = current_app_ils.document_indexer
         series_class = current_app_ils.series_record_cls
+        document_class = current_app_ils.document_record_cls
+
         self._validate_provider()
-        self.action = "none"
-        # finds the exact match, update records
-        self.document = self._match_document()
 
-        if self.document:
-            self.output_pid = self.document["pid"]
-            self.delete_records(self.document)
+        exact_match, partial_matches = self._match_document()
```

Comment:
In a future refactoring, this is also risky, same assumption that something happened before and maybe this `_match_document` will also set/change something in `self`.